### ISSUE TITLE
Unavailable Relay Due to Container Download Failures

### ISSUE THREAD

**Issue by northrup** [operable/cog#958]
There is the observed  behavior that a bundle service will become unavailable because the participating relay that was deployed is no longer available.  Upon inspection it is found that when the Cog relay server tried to refresh the container image for the target bundle relay the container registry did not respond within the appropriate time (or at all) and upon this failure the relay simply goes off line.

Would it be possible to have an configuration setting to disable such behaviors (and bubble up a soft error) rather than hard failing and effecting application performance?

**Comment by vanstee** [comment]
@northrup I was trying to reproduce this one. Could you post your relay logs while this is happening? Also are you saying Relay is crashing or just disconnecting from Cog when trying to enable a bundle? If Relay is still available but you're just not seeing the bundle enabled immediately after you run a command, you may want to check `$RELAY_COG_REFRESH_INTERVAL` and lower it if you're running commands before Relay has checked for new bundles to install (I think the default is 1m which is probably too long).

**Comment by northrup** [comment]
@vanstee the relay docker image is running, however that image itself should call the launching of a cast docker image that is pulled for twitter.  What happens is that when you invoke the twitter command it will say
```
Whoops! An error occurred. No Cog Relays supporting the `twitter` bundle are currently online
```
performing a `gogctl bundle disable twitter` followed by an enable will get it up and going for a few days, and then inevitably at some point it'll quit responding.

**Comment by vanstee** [comment]
@northrup Could you post your docker logs? If another process is deleting images out from under Relay that could potentially get Relay into a state where it thinks the image is available but docker doesn't have it (in which case Relay won't try to pull the image again).

**Comment by Shelton** [comment]
@vanstee Here are some notes from @northrup 

northrup [12:08 PM]  
So - I cannot find anywhere in the logs that would lead me to understand where any thing significant is happening.

shelton [12:08 PM]  
Hmm, interesting.

northrup [12:08 PM]  
I’m running this on a Digital Ocean droplet in their NYC3 data center

[12:09]  
and right now it doesn’t have a lot of activity… so it’ll go 24 hours or so without someone making a call against a runner

shelton [12:09 PM]  
Got it.

northrup [12:09 PM]  
however, when you then do - you get a “there’s no runner with that service"

**Comment by kevsmith** [comment]
This might be fixed in the upcoming 0.14.1 release. We found an issue where Relay would go into "zombie mode" -- still available but Cog didn't think it had any bundles installed -- if only some, but not all, of Relay's 3 message bus connections were disconnected.

**Comment by northrup** [comment]
@kevsmith interesting, what would cause the message bus disconnects? I have *a* relay that is running on the same server (same docker) as the postgres and cog images.

**Comment by kevsmith** [comment]
Depends on the environment. We saw this most recently w/someone running Cog on Amazon's ECS. Once every 24-ish hours the network would have a slight glitch and knock Relay partially offline.

You can tell if this is happening by looking at Cog and Relay's logs. You'll see a log entry from Cog like "Setting bundles list for Relay <relay_id>: []" with Relay log entries around the same time about being disconnected and reconnecting.

More details here: operable/cog#990

**Comment by kevsmith** [comment]
Is this still an issue? We believe this is fixed in 0.14.1.

**Comment by northrup** [comment]
This is no longer an issue, thanks!
